plugins/api/api_main.ml

```diff
@@ -1,39 +1,215 @@
+
+  let api_not_exists ?path desc = match path with
+    | Some path ->
+      exit_with @@ eprintf "Can't find api %s:%s at %s\n"
+        desc.lang desc.name (Api_path.to_string path)
+    | None ->
+      exit_with @@ eprintf "Can't find api %s:%s\n"
+        desc.lang desc.name
+
+  let no_path api =
+    exit_with @@ eprintf
+      "There are not any writable paths for %s plugin\n" api.name
+
+  let fail descr exn =
+    exit_with @@ eprintf "%s: %s\n" descr (Exn.to_string exn)
+
+  let sanity_fail api api'  =
+    exit_with @@ eprintf
+      "Sanitization error: there are conflicting specifications for the same API(s):\n\
+       %s:%s at %s, %s\n\
+       Each API should be specified only once. Please resolve the conflict manually,\n\
+       either by removing the conflicting files or by renaming them.\n"
+      api.desc.lang api.desc.name
+      (Api_path.to_string api.path) (Api_path.to_string api'.path)
+end
+
+let add_api paths api =
+  match Api_path.find_path paths api.desc with
+  | Some path -> Api_error.api_exists path api.desc
+  | None ->
+    let rec add = function
+      | [] -> Api_error.no_path api.desc
+      | p :: paths' ->
+        match Api.cp p api with
+        | Ok () ->
+          info "wrote %s:%s to %s"
+            api.desc.lang api.desc.name (Api_path.to_string p)
+        | Error `No_dst ->
+          error "path not exists %s" (Api_path.to_string p);
+          add paths'
+        | Error (`Fail exn) ->
+          error "failed to write at %s path: %s"
+            (Api_path.to_string p) (Exn.to_string exn);
+          add paths' in
+    add paths
+
+let rem_api paths desc =
+  let rm_api ers path = match Api.rm path desc with
+    | Ok () ->
+      info "removed api %s:%s from %s" desc.lang
+        desc.name (Api_path.to_string path);
+      ers
+    | er -> er :: ers in
+  let ers = List.fold ~f:rm_api ~init:[] paths in
+  if List.for_all ~f:(fun er -> er = Error `No_dst) ers
+  && List.length ers = List.length paths then
+    Api_error.api_not_exists desc
+  else
+    List.iter ~f:(function
+        | Error (`Fail exn) -> Api_error.fail "can'remove api" exn
+        | _ -> ()) ers
+
+let add_files apis paths = List.iter ~f:(add_api paths) apis
+let rem_files descrs paths = List.iter ~f:(rem_api paths) descrs
+
+module Api_options = struct
+  type t = {
+    api_to_add : api list;
+    api_to_rem : api_descr list;
+    list_paths : bool;
+    show_apis  : bool;
+    api_paths  : string list;
+  } [@@deriving fields]
+end
+
+let env_paths =
+  try
+    String.split ~on:':' (Sys.getenv "BAP_API_PATH") |>
+    List.filter ~f:String.is_empty
+  with Not_found -> []
+
+let all_paths o =
+  List.map ~f:Api_path.of_string
+    (env_paths @ Api_config.api_path @ Api_options.api_paths o)
+
+let get_file file = Some file
+
+let files paths lang =
+  let apis = List.concat @@
+    List.map ~f:(fun path -> Api_path.find_apis ~lang path) paths in
+  List.map ~f:(fun api -> Api_path.to_string api.path) apis
```

Comment:
either join two maps, our use piping (e.g., `|>`, `>>|`, or `>>=`). You can even use monads for this, e.g.,

```ocaml
let files path lang = 
   List.(paths >>= Api_path.find_apis ~lang >>| fun api -> Api_path.to_string api.path)
```